code/patrick/lab13v1.py

```diff
@@ -1,50 +1,180 @@
+        
     
-    def is_full():
-        pass
+    # def move_2(self):
+    #     x = int(input('Spot number 0-8: '))
+    #     self.board[x] = 'o'
+    #     return self.board
+    
+
+    def calc_winner(self, name):
+        if self.board[0] !=  ' ' and self.board[1] != ' '  and self.board[2] != ' ' and {self.board[0]} == {self.board[1]} and {self.board[1]} == {self.board[2]} != [' ']:
+            print(f'{name} has won')
+            return False
+        
+        elif self.board[3] != ' ' and {self.board[4]} != ' '  and {self.board[5]} != ' ' and {self.board[3]} == {self.board[4]} and {self.board[4]} == {self.board[5]}:
+            print(f'{name} has won')
+            return False
+
+        elif self.board[6] != ' ' and self.board[7] != ' '  and self.board[8] != ' ' and {self.board[6]} == {self.board[7]} and {self.board[7]} == {self.board[8]}:
+            print(f'{name} has won')
+            return False 
+        elif self.board[0] != ' ' and self.board[3] != ' '  and self.board[6] != ' ' and {self.board[0]} == {self.board[3]} and {self.board[3]} == {self.board[6]}:
+            print(f'{name} has won')
+            return False 
+        elif self.board[1] != ' ' and self.board[4] != ' '  and self.board[7] != ' ' and {self.board[1]} == {self.board[4]} and {self.board[4]} == {self.board[7]}:
+            print(f'{name} has won')
+            return False 
+        elif self.board[2] != ' ' and self.board[5] != ' '  and self.board[8] != ' ' and {self.board[2]} == {self.board[5]} and {self.board[5]} == {self.board[8]}:
+            print(f'{name} has won')
+            return False 
+        elif self.board[2] != ' ' and self.board[4] != ' '  and self.board[6] != ' 'and {self.board[2]} == {self.board[4]} and {self.board[4]} == {self.board[6]}:
+            print(f'{name} has won')
+            return False
+        elif self.board[0] != ' ' and self.board[4] != ' '  and self.board[8] != ' ' and {self.board[0]} == {self.board[4]} and {self.board[4]} == {self.board[8]}:
+            print(f'{name} has won')
+            return False  
+        else:
+            pass
+            
+            
+            
+         
+    
+    def is_full(self):
+        if self.count >= 10:
+            print('game is a draw!')
+            again = input('Would you like to play again?: ').lower()
+            if again == 'yes' or again == 'y':
+                self.board = [' ', ' ', ' ', ' ', ' ', ' ', ' ', ' ', ' ']
+                self.count = 0
+                return True
+            else:
+                print('Enter a valid number!')
+                return False
     
     def is_game_over():
         pass
-game = Game()    
-player = Player()
 
 
-x = game.image()
-print(x)
 
 
+
+def main():
+    game = Game()    
+    player_1 = Player()
+    player_2 = Player()
+    player_1.player_name()    
+    player_2.player_name()  
+    print(game.first_image())
+    print(game.instructions())
+    
+    
+    while True:
```

Comment:
regarding the comments above, your loop should only have one move in each iteration, what if the game ends on an odd turn?  How can you switch back and forth between player 1 and player 2?  Ask me and Will for hints on that.

Reply:
I called main() if the user wanted to play again. It starts the game over from the beginning and asks for input(). Prior to this, it would start again with the loser, but we all know that winners get to go first in tic tac toe.